juju/juju
# Macaroon v2

## Description of change

This PR starts the work to eliminate the use of macaroon.v2-unstable and bakerty.v2-unstable upstream repos. The work completed so far includes the use of macaroon.v2 everywhere and the use of bakery.v2 in the client (essentially bakery.v2/httpbakery).

The juju controller still uses bakery.v2-unstable to mint and store macaroons, and perform authorisation checks. The v2 APIs are totally different to v2-unstable so major work will be needed to get that fixed.

Much of the PR is the import changes to replace macaroon.v2-unstable with macaroon.v2 (and the similar bakery.v2 import changes).

There were several side effects that had to be addressed. We needed to move to using charmrepo.v4 instead of v3 in order to be compatible with the v2 macaroons etc. The v4 API changes had a flow on effect to the cmd/juju/application tests. The good part is that a huge chunk of code can be deleted (fakecharmstore) and all the deploy/bundle tests are much more consistent with each other.

Another side effect is that some code used juju/httprequest and other bits used httprequest.v1. Some work was needed to consolidate on httprequest.v1 and eliminate upstream conflicts.

To enable the use of macaroon.v2 everywhere, we need to depend on a fork I did of the bakery.v2-unstable repo. The fork introduced macaroon.v2 instead of v2-unstable to the bakery and did work to fix the juju/httprequest vw httprequest.v1 conflicts. As soon as we move to bakery.v2 everywhere, this temp dependency will be removed.

The end result is that all juju client code - CLI, api client, controller access to charm store etc - uses bakery.v2 and macaroon.v2 everywhere. Only the controller bakery and macaroon checks remain on bakery.v2 unstable.

## QA steps

We need to test old/new clients talking to new/old controllers, and also jaas.

Using a new juju client:
- login to jaas
- logout of jaas
- clear cookies from .local/share/juju
- log in to jaas again (both with -B and without -B to test interactive and browser auth)
- log in to an old 2.7 juju controller
- logout and see the change password prompt
- change the password, logout, and log in again
- add a user and use a different client home dir to register that user

Using an old 2.7 juju client
- perform the above checks against a new juju controller

Test CMR macaroon auth by:
- bootstrapping an old and a new controller
- in the old controller, create an offer
- add a user, grant that user access to the offer
- consume the offer in the new controller as the non admin user
- rinse and repeat where the offer is in the new controller
